Userland/Services/WebContent/BackingStoreManager.cpp

```diff
@@ -0,0 +1,139 @@
+#endif
+
+namespace WebContent {
+
+static bool s_iosurface_usage_disabled = false;
+
+void BackingStoreManager::disable_iosurface_usage()
+{
+    s_iosurface_usage_disabled = true;
+}
+
+#ifdef AK_OS_MACOS
+static Optional<Core::MachPort> s_browser_mach_port;
+void BackingStoreManager::set_browser_mach_port(Core::MachPort&& port)
+{
+    s_browser_mach_port = move(port);
+}
+#endif
+
+BackingStoreManager::BackingStoreManager(PageClient& page_client)
+    : m_page_client(page_client)
+{
+    m_backing_store_shrink_timer = Core::Timer::create_single_shot(3000, [this] {
+        resize_backing_stores_if_needed(WindowResizingInProgress::No);
+    });
+}
+
+void BackingStoreManager::restart_resize_timer()
+{
+    m_backing_store_shrink_timer->restart();
+}
+
+void BackingStoreManager::reallocate_backing_stores(Gfx::IntSize size)
+{
+#ifdef AK_OS_MACOS
+    if (!s_iosurface_usage_disabled) {
+        auto back_iosurface = Core::IOSurfaceHandle::create(size.width(), size.height());
+        auto back_iosurface_port = back_iosurface.create_mach_port(Core::MachPort::PortRight::Send);
+
+        auto front_iosurface = Core::IOSurfaceHandle::create(size.width(), size.height());
+        auto front_iosurface_port = front_iosurface.create_mach_port(Core::MachPort::PortRight::Send);
+
+        m_front_bitmap_id = m_next_bitmap_id++;
+        m_back_bitmap_id = m_next_bitmap_id++;
+
+        m_front_bitmap = Gfx::Bitmap::create_wrapper(Gfx::BitmapFormat::BGRA8888, size, size.width() * front_iosurface.bytes_per_element(), front_iosurface.data(), [handle = move(front_iosurface)] {}).release_value();
+        m_back_bitmap = Gfx::Bitmap::create_wrapper(Gfx::BitmapFormat::BGRA8888, size, size.width() * back_iosurface.bytes_per_element(), back_iosurface.data(), [handle = move(back_iosurface)] {}).release_value();
+
+        Core::Platform::BackingStoreMetadata metadata;
+        metadata.page_id = m_page_client.m_id;
+        metadata.front_backing_store_id = m_front_bitmap_id;
+        metadata.back_backing_store_id = m_back_bitmap_id;
+
+        Core::Platform::MessageWithBackingStores message;
+
+        message.header.msgh_remote_port = s_browser_mach_port->port();
+        message.header.msgh_local_port = MACH_PORT_NULL;
+        message.header.msgh_bits = MACH_MSGH_BITS(MACH_MSG_TYPE_COPY_SEND, 0) | MACH_MSGH_BITS_COMPLEX;
+        message.header.msgh_size = sizeof(message);
+        message.header.msgh_id = Core::Platform::BACKING_STORE_IOSURFACES_MESSAGE_ID;
+
+        message.body.msgh_descriptor_count = 2;
+
+        message.front_descriptor.name = front_iosurface_port.port();
+        message.front_descriptor.disposition = MACH_MSG_TYPE_COPY_SEND;
+        message.front_descriptor.type = MACH_MSG_PORT_DESCRIPTOR;
+
+        message.back_descriptor.name = back_iosurface_port.port();
+        message.back_descriptor.disposition = MACH_MSG_TYPE_COPY_SEND;
+        message.back_descriptor.type = MACH_MSG_PORT_DESCRIPTOR;
```

Comment:
You could even do

```suggestion
        message.back_descriptor.name = back_iosurface_port.release();
        message.back_descriptor.disposition = MACH_MSG_TYPE_MOVE_SEND;
        message.back_descriptor.type = MACH_MSG_PORT_DESCRIPTOR;
```

(and same above)

to avoid the extra `mach_port_deallocate()` syscalls in `~MachPort()` (and port refcount churn in the kernel).

Reply:
right, WebContent should not use an IOSurface port after sending it to the browser. then I should change `MachPort` to support MACH_MSG_TYPE_MOVE_SEND and make it do nothing in destructor for that type.

Reply:
hmm, MACH_MSG_TYPE_MOVE_SEND is a type, not right. I assume, if I change `disposition` to MACH_MSG_TYPE_MOVE_SEND I should not use `Core::MachPort` so `mach_port_deallocate` won't be invoked.

Reply:
You should use `.release()` as I said above, this will move the port right out of the `Core::MachPort` (replacing it with `MACH_PORT_NULL`), so the destructor won't do anything.

(Pardon my brevity, I'll be able to explain things in more detail later today)

Reply:
OK, so, there are the following types of port rights:
* __Dead name__ — what was previously a send or a send-once right to a port, but the port has since died (it's also possible to create dead names explicitly without there ever having been a corresponding port). This is like the writing end of a Unix pipe when the reading end has been closed;
* __Port set__ — groups/contains multiple receive rights;
* __Receive right__ — can receive from a port, and control it otherwise;
* __Send right__ — can send to the port;
* __Send-once right__ — can send to the port exactly once (sending consumes the right).

A single **name** (an integer, e.g. `(mach_port_t) 42`) can simultaneously denote several rights for a port (but it will always be the same port), for example a receive right and two send rights. (There are rules/limitations about this, mainly that send-once rights always have a single reference and don't share the name with any other right.) Send rights and dead names are refcounted; this is designed in a way that enables several parts of the same program to hold independent references to a port without knowing about each other or stepping on each other's toes. Speaking of which, you should never ever use `mach_port_destroy()` unless you're a Mach expert and know exactly what you're doing, because that ignores the refcounting and deallocates _all_ the rights a name refers to, which other parts of the program that might be independently using the same right might not appreciate.

In C, the `mach_port_right_t` enum (`MACH_PORT_RIGHT_RECEIVE`, `MACH_PORT_RIGHT_PORT_SET`, ...) defines individual port right types. This is what you'd pass to `mach_port_allocate()` and `mach_port_{get,mod}_refs()`. This is also what `Core::MachPort::PortRight` mirrors.

The `mach_port_type_t` type is a bitmask of `MACH_PORT_TYPE_RECEIVE`, `MACH_PORT_TYPE_SEND` (note `TYPE` vs `RIGHT` before) bits, describing what rights a port name (in some IPC namespace) denotes, 'cause it can be multiple at the same time. You'd get this as output from `mach_port_names()` and `mach_port_type()` queries.

Then, there is the **message type**, also called **disposition** in untyped IPC variant of Mach. This describes how a port right is _sent in a message_. In typed IPC (as used in GNU Mach) these are used to describe every field in a Mach message, so there are `MACH_MSG_TYPE_CHAR`, `MACH_MSG_TYPE_INTEGER_32`, `MACH_MSG_TYPE_STRING`, and so on.

With untyped IPC (as implemented in OSF's version of Mach, from which Darwin's version of Mach derives), plain data fields don't need `MACH_MSG_TYPE`s assigned to them — instead, as you've already seen, you just kind of stuff a chunk of memory into a message, and the kernel copies it as-is, and you're left dealing with potential ABI differences (such as byte order when messaging between i386 and PowerPC processes, or sizes on 32-bit vs 64-bit) by yourself (MIG does it for you if you use MIG).

Even with untyped IPC, you still have to specify message types / dispositions for the ports you transfer in a message. There isn't just a "`MACH_MSG_TYPE_PORT`"; there are many port message types that describe how a port right that the message will carry is to be derived from a port right that you hold. `MACH_MSG_TYPE_MOVE_RECEIVE`, `MACH_MSG_TYPE_MOVE_SEND`, and `MACH_MSG_TYPE_MOVE_SEND_ONCE` _move_ a right that you have into the message, so you no longer have the right after the message is sent (the name becomes free to be reused for something else if it denoted no other rights). `MACH_MSG_TYPE_MAKE_SEND` and `MACH_MSG_TYPE_MAKE_SEND_ONCE` _make_ (and place into the message) a new send/send-once right from a receive right that you hold; you still hold the receive right after the message is sent. `MACH_MSG_TYPE_COPY_SEND` _copies_ a send right that you hold; you still hold the send right after the message is sent (contrast with `MACH_MSG_TYPE_MOVE_SEND`).

There are also `MACH_MSG_TYPE_PORT_RECEIVE`, `MACH_MSG_TYPE_PORT_SEND`, and `MACH_MSG_TYPE_PORT_SEND_ONCE` (and in GNU Mach, `MACH_MSG_TYPE_PROTECTED_PAYLOAD`, which deserves its own essay). These are basically aliases for "move" variants, so `MACH_MSG_TYPE_MOVE_RECEIVE` and so on. These are intended to be used when dissecting a received message (as opposed to sending one), since in a received message, all contained port right appear to have been moved, no matter which actual disposition has been used by the sender, since the receiver should only care about the type of port right contained in the message, not how the sender created it.

In MIG, `mach_port_t`, `mach_port_send_t`, `mach_port_make_send_t`, etc. are all separate types. They all map to `mach_port_t` in C, but cause MIG to use different dispositions when sending the message.

---

Finally, let's look at two curious routines, `mach_port_insert_right()` and `mach_port_extract_right()`. In C, their signatures are:

```c
kern_return_t mach_port_insert_right(ipc_space_t task, mach_port_t name, mach_port_t right, mach_msg_type_name_t right_type);
kern_return_t mach_port_extract_right(ipc_space_t task, mach_port_t name, mach_msg_type_name_t desired_type, mach_port_t *right, mach_msg_type_name_t *acquired_type);
```

(`ipc_space_t` = `task_t` = `mach_port_t` in userland, this is just a (send) right to the task port.) You'd use `mach_port_insert_right` like this:

```c
// We hold a receive right to a port, under the name "port".
// Insert into target_task, under the name 42, a send right for the same port.
err = mach_port_insert_right(target_task, 42, port, MACH_MSG_TYPE_MAKE_SEND);
```

So it seems that `mach_port_insert_right()` is reusing the `MACH_MSG_TYPE` constants to describe how to derive the new right (that's to be inserted into the target task) from the existing right. But if you look at the actual MIG signature:

```mig
routine mach_port_insert_right(
		task		: ipc_space_t;
		name		: mach_port_name_t;
		poly		: mach_port_poly_t);
```

you can see that it only actually takes three arguments! It's `mach_port_poly_t` type that generates the extra `mach_msg_type_name_t` argument in C; basically, you're telling MIG what you want it to set the disposition of `poly` right to when sending the message. The RPC request message, as sent to the kernel, will just contain `task`, `name`, and `poly`, and the `mach_port_insert_right()` implementation in the kernel will just insert `poly` as contained in the message into the target task.

But, the same doesn't apply to `mach_port_extract_right`:

```mig
routine mach_port_extract_right(
		task		: ipc_space_t;
		name		: mach_port_name_t;
		msgt_name	: mach_msg_type_name_t;
	out	poly		: mach_port_poly_t);
```

Here, `mach_msg_type_name_t *acquired_type` is what the `mach_port_poly_t` generates, but `desired_type` / `msgt_name` is an explicit argument that's actually contained in the RPC request message, and is explicitly processed by the `mach_port_extract_right()` implementation.

---

I've previously written about Mach IPC [in Darling docs](https://docs.darlinghq.org/internals/macos-specifics/mach-ports.html) and on typed vs untyped IPC [on the Hurd mailing list](https://mail.gnu.org/archive/html/bug-hurd/2023-02/msg00157.html), you might find those helpful.